core/trino-main/src/main/java/io/trino/execution/buffer/OutputBufferStateMachine.java

```diff
@@ -70,4 +75,17 @@ public boolean abort()
     {
         return state.setIf(ABORTED, oldState -> !oldState.isTerminal());
     }
+
+    public boolean fail(Throwable throwable)
+    {
+        requireNonNull(throwable, "throwable is null");
+
+        failureCause.compareAndSet(null, throwable);
```

Comment:
very that `failureCause` is not overwritten (it is null)?

Reply:
This method is allowed to be called multiple times (similar to how it is implemented in other state machines). The contract is that the method has to preserve only the first failure that made the transition.

Reply:
The code seems to be self explanatory and aligns with what is done in other state machines. How strongly do you feel about having an explicit comment here?